src/code/z_play.c

```diff
@@ -74,39 +74,206 @@ s32 Play_InCsMode(PlayState* this) {
+void Play_ClearAllSubCameras(PlayState* this) {
+    s16 subCamId;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80169A50.s")
+    for (subCamId = CAM_ID_SUB_FIRST; subCamId < NUM_CAMS; subCamId++) {
+        if (this->cameraPtrs[subCamId] != NULL) {
+            Play_ClearCamera(this, subCamId);
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/Play_CameraChangeSetting.s")
+    this->activeCamId = CAM_ID_MAIN;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80169AFC.s")
+Camera* Play_GetCamera(PlayState* this, s16 camId) {
+    s16 camIdx = (camId == CAM_ID_NONE) ? this->activeCamId : camId;
+
+    return this->cameraPtrs[camIdx];
+}
+
+s32 Play_SetCameraAtEye(PlayState* this, s16 camId, Vec3f* at, Vec3f* eye) {
+    s32 successfullySet = 0;
+    s16 camIdx = (camId == CAM_ID_NONE) ? this->activeCamId : camId;
+    Camera* camera = this->cameraPtrs[camIdx];
+
+    successfullySet |= Camera_SetViewParam(camera, CAM_VIEW_AT, at);
+    successfullySet <<= 1;
+    successfullySet |= Camera_SetViewParam(camera, CAM_VIEW_EYE, eye);
+
+    camera->dist = Math3D_Distance(at, eye);
+
+    if (camera->focalActor != NULL) {
+        camera->atActorOffset.x = at->x - camera->focalActor->world.pos.x;
+        camera->atActorOffset.y = at->y - camera->focalActor->world.pos.y;
+        camera->atActorOffset.z = at->z - camera->focalActor->world.pos.z;
+    } else {
+        camera->atActorOffset.x = camera->atActorOffset.y = camera->atActorOffset.z = 0.0f;
+    }
+
+    camera->atLerpStepScale = 0.01f;
+
+    return successfullySet;
+}
+
+s32 Play_SetCameraAtEyeUp(PlayState* this, s16 camId, Vec3f* at, Vec3f* eye, Vec3f* up) {
+    s32 successfullySet = 0;
+    s16 camIdx = (camId == CAM_ID_NONE) ? this->activeCamId : camId;
+    Camera* camera = this->cameraPtrs[camIdx];
+
+    successfullySet |= Camera_SetViewParam(camera, CAM_VIEW_AT, at);
+    successfullySet <<= 1;
+    successfullySet |= Camera_SetViewParam(camera, CAM_VIEW_EYE, eye);
+    successfullySet <<= 1;
+    successfullySet |= Camera_SetViewParam(camera, CAM_VIEW_UP, up);
+
+    camera->dist = Math3D_Distance(at, eye);
+
+    if (camera->focalActor != NULL) {
+        camera->atActorOffset.x = at->x - camera->focalActor->world.pos.x;
+        camera->atActorOffset.y = at->y - camera->focalActor->world.pos.y;
+        camera->atActorOffset.z = at->z - camera->focalActor->world.pos.z;
+    } else {
+        camera->atActorOffset.x = camera->atActorOffset.y = camera->atActorOffset.z = 0.0f;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/Play_CameraGetUID.s")
+    camera->atLerpStepScale = 0.01f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_play/func_80169BF8.s")
+    return successfullySet;
+}
+
+s32 Play_SetCameraFov(PlayState* this, s16 camId, f32 fov) {
+    s32 successfullySet = Camera_SetViewParam(this->cameraPtrs[camId], CAM_VIEW_FOV, &fov) & 1;
+
+    if (1) {}
+    return successfullySet;
```

Comment:
I suppose this definitely requires a temp?

Reply:
https://decomp.me/scratch/Jk3Mz

Reply:
have you tried passing GameState instead and recasting to remove the if(1) thing? it can't be tried on that scratch, i get random enconding errors

Reply:
The issue is that `cameraPtrs` isn't part of `GameState`, that's specifically part of the `PlayState` struct. So it would have to be  casted to `PlayState`, which doesn't seem to change anything:
https://decomp.me/scratch/aAgkf